content/articles/csrf-protection-in-php/index.md

```diff
@@ -0,0 +1,437 @@
+        if (empty($this->session[$this->sessionTokenLabel])) {
+            $this->session[$this->sessionTokenLabel] = bin2hex(openssl_random_pseudo_bytes(32));
+        }
+
+        if ($this->hmac_ip !== false) {
+            $token = $this->hMacWithIp($this->session[$this->sessionTokenLabel]);
+        } else {
+            $token = $this->session[$this->sessionTokenLabel];
+        }
+        return $token;
+    }
+```
+
+### Step 2: Render contact form with CSRF token
+This is a contact form HTML with the usual fields name, email, subject, and message. Added to that is the hidden field csrf-token with the generated token.
+
+The submit action processes jQuery form validation before posting the parameters to the PHP.
+
+The client-side validation script handles the basic validation on submit. It applies the not-empty check on each field.
+
+```php
+index.php (HTML Template)
+
+<html>
+<head>
+<title>CSRF Protection using PHP</title>
+<link rel="stylesheet" type="text/css"
+    href="assets/css/style.css" />
+<script src="vendor/jquery/jquery-3.2.1.min.js"></script>
+</head>
+<body>
+    <div class="container">
+        <h1>CSRF Protection using PHP</h1>
+        <form name="frmContact" id="cnt-frm" class="form"
+            frmContact"" method="post" action="" enctype="multipart/form-data"
+            onsubmit="return validateContactForm()">
+
+            <div class="-row">
+                <div class="label">
+                    Name <span id="userName-info" class="validation-message"></span>
+                </div>
+                <input type="text" class="input" name="userName"
+                    id="userName"
+                    value="<?php if(!empty($_POST['userName'])&& $type == 'error'){ echo $_POST['userName'];}?>" />
+            </div>
+            <div class="-row">
+                <div class="label">
+                    Email <span id="userEmail-info" class="validation-message"></span>
+                </div>
+                <input type="text" class="input" name="userEmail"
+                    id="userEmail"
+                    value="<?php if(!empty($_POST['userEmail'])&& $type == 'error'){ echo $_POST['userEmail'];}?>" />
+            </div>
+            <div class="-row">
+                <div class="label">
+                    Subject <span id="subject-info" class="validation-message"></span>
+                </div>
+                <input type="text" class="input" name="subject" id="subject"
+                    value="<?php if(!empty($_POST['subject'])&& $type == 'error'){ echo $_POST['subject'];}?>" />
+            </div>
+            <div class="-row">
+                <div class="label">
+                    Message <span id="userMessage-info" class="validation-message"></span>
+                </div>
+                <textarea name="content" id="content" class="-input" cols="60"
+                    rows="6"><?php if(!empty($_POST['content'])&& $type == 'error'){ echo $_POST['content'];}?></textarea>
```

Comment:
Formatting error in code. Please check if the code runs, as this indicates there exists errors in the code.